sdk/src/main/scala/com/horizen/block/SidechainsHashMap.scala

```diff
@@ -33,97 +33,117 @@ class SidechainHashList
+  }
 }
 
-class SidechainHashMap
+class SidechainsHashMap
 {
 
-  val sidechainHashMap: mutable.Map[ByteArrayWrapper, SidechainHashList] = new mutable.HashMap[ByteArrayWrapper, SidechainHashList]()
+  val sidechainsHashMap: mutable.Map[ByteArrayWrapper, SidechainHashList] = new mutable.HashMap[ByteArrayWrapper, SidechainHashList]()
 
   def addTransactionOutputs(sidechainId: ByteArrayWrapper,
-                            outputs: Seq[SidechainRelatedMainchainOutput[_ <: Box[_ <: Proposition]]]): SidechainHashMap = {
-    sidechainHashMap.get(sidechainId) match {
+                            outputs: Seq[SidechainRelatedMainchainOutput[_ <: Box[_ <: Proposition]]]): Unit = {
+    sidechainsHashMap.get(sidechainId) match {
       case Some(shl) => shl.addTransactionHash(outputs.map(_.hash()))
       case None =>
         val shl = new SidechainHashList()
         shl.addTransactionHash(outputs.map(_.hash()))
-        sidechainHashMap.put(sidechainId, shl)
+        sidechainsHashMap.put(sidechainId, shl)
     }
-
-    this
   }
 
-  def addCertificate(certificate: MainchainBackwardTransferCertificate): SidechainHashMap = {
+  def addCertificate(certificate: MainchainBackwardTransferCertificate): Unit = {
     val sidechainId = new ByteArrayWrapper(certificate.sidechainId)
     val certificateHash =
         BytesUtils.reverseBytes(
           Utils.doubleSHA256Hash(
             certificate.certificateBytes
           )
         )
-    sidechainHashMap.get(sidechainId) match {
+    sidechainsHashMap.get(sidechainId) match {
       case Some(shl) =>shl.addWithdrawalCertificateHash(certificateHash)
       case None =>
         val shl = new SidechainHashList()
         shl.addWithdrawalCertificateHash(certificateHash)
-        sidechainHashMap.put(sidechainId, shl)
+        sidechainsHashMap.put(sidechainId, shl)
     }
-
-    this
   }
 
-  def getMerkleRoot(sidechainId: ByteArrayWrapper): Array[Byte] = {
-    sidechainHashMap.get(sidechainId) match {
+  def getSidechainHash(sidechainId: ByteArrayWrapper): Array[Byte] = {
+    sidechainsHashMap.get(sidechainId) match {
       case Some(shl) => shl.getSidechainHash(sidechainId.data)
       case None => Array.emptyByteArray
     }
   }
 
-  def getMerkleRoot(sidechainId: Array[Byte]): Array[Byte] =
-    getMerkleRoot(new ByteArrayWrapper(sidechainId))
+  private def getMerkleTree: MerkleTree = {
+    val merkleTreeLeaves = sidechainsHashMap.toSeq.sortWith(_._1 < _._1)
+      .map(pair => {
+        getSidechainHash(pair._1)
+      })
+
+    MerkleTree.createMerkleTree(merkleTreeLeaves.asJava)
+  }
 
-  def getMerkleTree: MerkleTree = {
-    val merkleTreeLeaves = sidechainHashMap.toSeq.sortWith(_._1 < _._1)
+  private def getFullMerkleTree: MerkleTree = {
+    var merkleTreeLeaves = sidechainsHashMap.toSeq.sortWith(_._1 < _._1)
       .map(pair => {
-        getMerkleRoot(pair._1)
+        getSidechainHash(pair._1)
       })
 
+    var fullMktSize = 2
```

Comment:
Probably start from `1`. In current implementation of MerkleTree, in case of single leaf - merkle root is equal to that leaf.